### PR TITLE
WIP osx support

### PR DESCRIPTION
**Pull request by sharpobject**
I just messed with things until it built successfully. These changes might not actually work, because I haven't tried running any games with them, but I wanted to put this out there to see what kind of non-correctness-related changes it needs.

Note that to actually build on OSX you will need these changes plus the changes in #16. I did not include the changes in #16 because that would be annoying to read.

### CODE REVIEW

File: src/lib/ggpo/network/udp.cpp [hunk 1]
```diff
@@ -15,10 +15,13 @@ CreateSocket(int bind_port, int retries)
    sockaddr_in sin;
    int port;
    int optval = 1;
+   struct linger loptval;
+   loptval.l_onoff = 0;
+   loptval.l_linger = 0;
 
    s = socket(AF_INET, SOCK_DGRAM, 0);
    setsockopt(s, SOL_SOCKET, SO_REUSEADDR, (const char *)&optval, sizeof optval);
-   setsockopt(s, SOL_SOCKET, SO_DONTLINGER, (const char *)&optval, sizeof optval);
```
Comment (sharpobject) [inline src/lib/ggpo/network/udp.cpp]: I have a question about the intended and actual behavior of this line. The docs at https://docs.microsoft.com/en-us/windows/win32/api/winsock/ns-winsock-linger say that calling setsockopt with `SO_DONTLINGER` and optval 1 is equivalent to calling setsockopt with `SO_LINGER` and optval `struct linger{1, 1}`. This seems obviously wrong though, like turning ON the DONTLINGER setting should turn OFF the LINGER setting.....

So, you do want to turn off LINGER, right?

Comment (sharpobject) [reply src/lib/ggpo/network/udp.cpp]: Well, the docs at https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-setsockopt say the opposite, so I'll just assume it's a problem with the docs....

**Comment by pond3r** [comment]
I appreciate the effort, but am a little concerned with the approach.  Rather than making the Windows API the authoritative version, I would prefer it if we created a platform abstraction layer and made the rest of the code depend on it.  For example, instead of mapping a bunch of Window API concepts to their unix equivalents and making ggpo code to the Windows API, we should define a very narrow platform abstraction layer and implement it on both Windows and Unix.  For the most part, I think we can rely on posix or posix-like standards (e.g. to get the definition of `socklen_t` on Windows, I think we just need to include `WS2tcpip.h`).

I can take a stab at what I think this looks like in a few days if you can hold off until then.

**Comment by sharpobject** [comment]
Alright. That sounds good to me!

I don't really know what's going on in the Windows events code, and it seems like it really is substantially easier to use one of the existing "windows events on posix" things than rewrite it. Everything but Windows events could be much cleaner than what I've written though, with not much effort.

**Comment by pond3r** [comment]
Hey.  I've started a branch that demonstrates the way I'd like to do cross-platform support.  Feel free to check it out here:

https://github.com/pond3r/ggpo/tree/dev/crossplatform

I think the right way to handle this is to make a `platform_osx.h` file and put most of the implementation there (and a `.cpp` if you need it).  Would you be willing to stage a new pull request into `dev/crossplatform`?

Thanks!

**Pull request closed without merging**